### PR TITLE
Reorganising output and out opcodes revised 

### PR DESCRIPTION
**Pull request by vlazzarini**
This PR modifies the output spout so that each instrument can hold its own separate output.
It also revises all output opcodes to use a single function.

Current design in Csound 6
====

1. out opcodes etc -> mix into csound->spraw (non interleaved)
2. at every k-cycle make_interleave() interleaves spoutraw into csound->spout
3. locks in each out opcode to protect writing into csound->spraw
4. output is implemented by various opcodes.

Changes in this PR
====

1. all output opcodes share the same inner loop code.
2. locks removed
3. interleaving takes place at
a more reasonable place 

 
PROS: 
====
 - fixes issue #1818 (but needs testing)
 - this paves the ground for the implementation of oversampling in instruments.
 - out opcodes refactored should
facilitate maintenance. 

PR is now ready.

### CODE REVIEW

**Comment by vlazzarini** [comment]
Revised this now, removed the per-instance spout because of unnecessary extra copying. This now is performing as well as the code in develop.

There's an issue with PARCS, where the mixing of various thread outputs is not right. I am trying to figure out what is going on. Probably I have the wrong assumptions.

**Comment by vlazzarini** [comment]
PARCS restored and runs slightly faster than csound 6.18.

There's an oddity with PARCS in Csound 7 that it gives louder amplitudes at the end of some sections, but this varies from run to run. It's happening here and in the develop branch code. I suspect it is to do with the new parser (and how certain things were not put back into it).

As far as the output goes, it is all working correctly now.
Other things need to be tested, but I think this design is good.

**Comment by jpffitch** [comment]
Sorry not to put this in a review but I cannot find how to do that.
Seems a step forward and improves efficiency.  I have not read all the code , but I trust you to get it basically right.

Extra thought; not all instruments call "out" in any of its forms, so tfor his class can skip this admin code and avoid memory allocation for instance local spout.  The semantic analysis should be able to flag to which class an instrument belongs.  Coud be done oither ways of course.

You should merge this PR

**Comment by vlazzarini** [comment]
thanks, I'll do the tests and if they pass I'll report back and see if we can merge.

In the end the local spout was the wrong idea. It just introduced an extra copying layer that slowed the code down.

But I think the other changes are good.

**Comment by vlazzarini** [comment]
@jpffitch @kunstmusik this is ready now for review. I have tested the code and it is working with all out opcodes.

**Comment by vlazzarini** [comment]
Can we merge this?

**Comment by kunstmusik** [comment]
@vlazzarini Apologies for delay, will review this evening!

**Comment by vlazzarini** [comment]
thanks

**Comment by vlazzarini** [comment]
I've updated it with some outstanding work I had thought of in the meantime.

File: Engine/musmon.c [hunk 1]
```diff
@@ -270,7 +270,10 @@ int musmon(CSOUND *csound)
     csound->nspout = csound->ksmps * csound->nchnls;  /* alloc spin & spout */
     csound->nspin = csound->ksmps * csound->inchnls; /* JPff: in preparation */
     csound->spin  = (MYFLT *) csound->Calloc(csound, csound->nspin*sizeof(MYFLT));
-    csound->spraw = (MYFLT *) csound->Calloc(csound, csound->nspout*sizeof(MYFLT));
+    /* VL: 27.1.24 spraw redeployed as thread buffer */
+    printf("threads: %d \n", csound->oparms->numThreads);
```
Comment (kunstmusik) [inline Engine/musmon.c]: This looks like a debug message for testing? If so, suggest to remove.

File: Engine/insert.c [hunk 2]
```diff
@@ -1881,14 +1881,14 @@ int subinstr(CSOUND *csound, SUBINST *p)
     return csoundPerfError(csound, &(p->h),
                            Str("subinstr: not initialised"));
   }
+  // This not needed since spout buffer is already managed by
```
Comment (kunstmusik) [inline Engine/insert.c]: Looking at this comment, does it mean the code below should be removed?

File: Engine/insert.c [hunk 3]
```diff
@@ -2668,7 +2668,8 @@ static void instance(CSOUND *csound, int insno)
     var->memBlock = (CS_VAR_MEM*)(temp - CS_VAR_TYPE_OFFSET);
     var->memBlock->value = csound->ekr;
   }
-
+  
+  //printf("alloc spout: %p \n", ip->spout);
```
Comment (kunstmusik) [inline Engine/insert.c]: Remove dead code

File: Top/csound.c [hunk 4]
```diff
@@ -1767,10 +1741,12 @@ int kperf_nodebug(CSOUND *csound)
     /* for one kcnt: */
     if (csound->oparms_.sfread)         /*   if audio_infile open  */
       csound->spinrecv(csound);         /*      fill the spin buf  */
-    csound->spoutactive = 0;            /*   make spout inactive   */
     /* clear spout */
     memset(csound->spout, 0, csound->nspout*sizeof(MYFLT));
-    memset(csound->spraw, 0, csound->nspout*sizeof(MYFLT));
+    memset(csound->spraw,0,
+           sizeof(MYFLT)*csound->nspout*csound->oparms->numThreads);
+    // 27.1.24 spraw is not used anymore
```
Comment (kunstmusik) [inline Top/csound.c]: Seems like this comment is a bit confusing about spraw not used anymore? Looks like it is still in use. What do you think about giving spraw a different name, or at least adding comments to the field in csoundCore.h? I suspect future devs would appreciate some documentation about the spout vs spraw.

File: include/csoundCore.h [hunk 5]
```diff
@@ -582,6 +582,7 @@ typedef struct CORFIL {
                               (used by opcodes) */
     MYFLT   *spin;         /* offset into csound->spin */
     MYFLT   *spout;        /* offset into csound->spout, or local spout */
+    int      spout_flag;
```
Comment (kunstmusik) [inline include/csoundCore.h]: Is this necessary? I did a search through the code and found spout_flag was only ever set to 0 (unless I'm missing something).

**Comment by kunstmusik** [comment]
@vlazzarini Things looked good as far as I could see though I left some comments about possible dead code and tidying matters.

**Comment by vlazzarini** [comment]
Thanks I think some of the comments are not applicable anymore, as I made some later changes. spraw was removed and them later put back so that's one. I'll review today

File: Engine/musmon.c (hunk 1 above)
Comment (vlazzarini) [reply Engine/musmon.c]: done.

File: Engine/insert.c (hunk 2 above)
Comment (vlazzarini) [reply Engine/insert.c]: spurious comment left over from earlier changes. removed it

File: Engine/insert.c (hunk 3 above)
Comment (vlazzarini) [reply Engine/insert.c]: done

File: Top/csound.c (hunk 4 above)
Comment (vlazzarini) [reply Top/csound.c]: I'll think of a name change. I removed it earlier but then restored it.

File: include/csoundCore.h (hunk 5 above)
Comment (vlazzarini) [reply include/csoundCore.h]: removed.

**Comment by vlazzarini** [comment]
Ready now.

**Comment by kunstmusik** [comment]
Looks good to me, merging now, thanks!

**Pull request merged**